File: web/js/appearance.js
/**
 * Appearance engine — themes, typography, layout, and writing prefs.
 * Applies CSS data-attributes immediately, persists to localStorage,
 * and syncs to the desktop JSON file when Eel is available.
 */

const STORAGE_KEY = 'kosistenz-appearance';

export const COLOR_SLOTS = [
    { id: 'pageBg', label: 'Page background' },
    { id: 'widgetBg', label: 'Widget background' },
    { id: 'widgetBorder', label: 'Widget borders' },
    { id: 'titles', label: 'Titles' },
    { id: 'accent', label: 'Accent' },
    { id: 'done', label: 'Done' },
    { id: 'openNext', label: 'Open / next' },
    { id: 'sidebar', label: 'Sidebar' },
];

export const INK_LIGHT = '#f7fafc';
export const INK_DARK = '#1a1814';

export const THEME_PALETTES = {
    ocean: {
        pageBg: '#121c26',
        widgetBg: '#1d2c3b',
        widgetBorder: '#2c3d4e',
        titles: '#eef3f7',
        accent: '#4f8fcf',
        done: '#5ebb8e',
        openNext: '#d4a054',
        sidebar: '#0e1620',
    },
    midnight: {
        pageBg: '#0e0d0b',
        widgetBg: '#161512',
        widgetBorder: '#2a2722',
        titles: '#f0eee9',
        accent: '#e0b355',
        done: '#5ebb8e',
        openNext: '#e0b355',
        sidebar: '#161512',
    },
    slate: {
        pageBg: '#171e2b',
        widgetBg: '#243044',
        widgetBorder: '#2c3848',
        titles: '#eef2f6',
        accent: '#4f8fcf',
        done: '#5ebb8e',
        openNext: '#d4a054',
        sidebar: '#121824',
    },
    paper: {
        pageBg: '#f7f3eb',
        widgetBg: '#f1ebe0',
        widgetBorder: '#d8d0c2',
        titles: '#1b1814',
        accent: '#4f8fcf',
        done: '#2f7d57',
        openNext: '#b5791f',
        sidebar: '#f3eee4',
    },
    forest: {
        pageBg: '#141e1a',
        widgetBg: '#20312b',
        widgetBorder: '#2a3c36',
        titles: '#eef4f0',
        accent: '#4f8fcf',
        done: '#6bc49a',
        openNext: '#d4a054',
        sidebar: '#101816',
    },
    dusk: {
        pageBg: '#1b1824',
        widgetBg: '#2b2738',
        widgetBorder: '#3a3548',
        titles: '#f2eef6',
        accent: '#4f8fcf',
        done: '#5ebb8e',
        openNext: '#d4a054',
        sidebar: '#16131e',
    },
};

export const DEFAULTS = {
    theme: 'ocean',
    accent: 'sky',
    customAccent: '#4F8FCF',
    font: 'system',
    density: 'comfortable',
    radius: 'soft',
    width: 'standard',
    sidebar: 'compact',
    todayLayout: 'split',
    todayOrder: 'todo,workout,journal',
    todayTodo: true,
    todayWorkout: true,
    todayJournal: true,
    journalFontSize: 17,
    timerMinutes: 10,
    autoFocus: false,
    reducedMotion: false,
    highContrast: false,
    colorOverrides: {},
    widgetBorderWidth: 1,
    inkAuto: true,
    ink: '',
    activePresetId: '',
    userPresets: [],
};

export const THEMES = [
    { id: 'ocean', label: 'Ocean' },
    { id: 'midnight', label: 'Midnight' },
    { id: 'slate', label: 'Slate' },
    { id: 'paper', label: 'Paper' },
    { id: 'forest', label: 'Forest' },
    { id: 'dusk', label: 'Dusk' },
];

export const ACCENTS = [
    { id: 'sky', label: 'Blue', hex: '#4F8FCF' },
    { id: 'teal', label: 'Teal', hex: '#2A9A8C' },
    { id: 'amber', label: 'Ochre', hex: '#C8892C' },
    { id: 'rose', label: 'Rose', hex: '#C45C6A' },
    { id: 'violet', label: 'Violet', hex: '#7A6CB5' },
    { id: 'lime', label: 'Moss', hex: '#6A9A6E' },
    { id: 'custom', label: 'Custom', hex: null },
];

let current = { ...DEFAULTS, colorOverrides: {}, userPresets: [] };
const listeners = new Set();

function readLocal() {
    try {
        const raw = localStorage.getItem(STORAGE_KEY);
        if (!raw) return {};
        const parsed = JSON.parse(raw);
        return parsed && typeof parsed === 'object' ? parsed : {};
    } catch (_) {
        return {};
    }
}

function writeLocal(settings) {
    try {
        localStorage.setItem(STORAGE_KEY, JSON.stringify(settings));
    } catch (_) {
        /* private mode / quota */
    }
}

export function normalizeHex(hex, fallback = '') {
    let h = String(hex || '').trim();
    if (!h) return fallback;
    if (h[0] !== '#') h = `#${h}`;
    if (h.length === 4 && /^#[0-9a-fA-F]{3}$/.test(h)) {
        h = `#${h[1]}${h[1]}${h[2]}${h[2]}${h[3]}${h[3]}`;
    }
    if (!/^#[0-9a-fA-F]{6}$/.test(h)) return fallback;
    return h.toLowerCase();
}

function hexToHsl(hex) {
    const normalized = normalizeHex(hex, '#4f8fcf').slice(1);
    const r = parseInt(normalized.slice(0, 2), 16) / 255;
    const g = parseInt(normalized.slice(2, 4), 16) / 255;
    const b = parseInt(normalized.slice(4, 6), 16) / 255;
    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);
    let hue = 0;
    let sat = 0;
    const light = (max + min) / 2;
    const d = max - min;
    if (d !== 0) {
        sat = d / (1 - Math.abs(2 * light - 1));
        switch (max) {
            case r:
                hue = ((g - b) / d) % 6;
                break;
            case g:
                hue = (b - r) / d + 2;
                break;
            default:
                hue = (r - g) / d + 4;
        }
        hue *= 60;
        if (hue < 0) hue += 360;
    }
    return {
        h: Math.round(hue),
        s: Math.round(sat * 100),
        l: Math.round(light * 100),
    };
}

function srgbToLinear(channel) {
    return channel <= 0.04045 ? channel / 12.92 : ((channel + 0.055) / 1.055) ** 2.4;
}

export function relativeLuminance(hex) {
    const n = normalizeHex(hex, '');
    if (!n) return 0;
    const r = parseInt(n.slice(1, 3), 16) / 255;
    const g = parseInt(n.slice(3, 5), 16) / 255;
    const b = parseInt(n.slice(5, 7), 16) / 255;
    return 0.2126 * srgbToLinear(r) + 0.7152 * srgbToLinear(g) + 0.0722 * srgbToLinear(b);
}

export function inkForHex(hex) {
    return relativeLuminance(hex) > 0.45 ? INK_DARK : INK_LIGHT;
}

export function paletteFor(theme) {
    return { ...(THEME_PALETTES[theme] || THEME_PALETTES.ocean) };
}

export function resolveAccentHex(settings) {
    const overrides = settings.colorOverrides || {};
    const fromOverride = normalizeHex(overrides.accent, '');
    if (fromOverride) return fromOverride;
    if (settings.accent === 'custom') {
        const custom = normalizeHex(settings.customAccent, '');
        if (custom) return custom;
    }
    const preset = ACCENTS.find((a) => a.id === settings.accent);
    if (preset?.hex) return normalizeHex(preset.hex, paletteFor(settings.theme).accent);
    return paletteFor(settings.theme).accent;
}

export function resolveColors(settings) {
    const out = paletteFor(settings.theme);
    const overrides = settings.colorOverrides || {};
    COLOR_SLOTS.forEach(({ id }) => {
        if (id === 'accent') return;
        const hx = normalizeHex(overrides[id], '');
        if (hx) out[id] = hx;
    });
    out.accent = resolveAccentHex(settings);
    return out;
}

export function resolveInk(settings) {
    if (settings.inkAuto !== false) return inkForHex(resolveAccentHex(settings));
    return normalizeHex(settings.ink, '') || inkForHex(resolveAccentHex(settings));
}

function mergeSettings(partial) {
    const next = {
        ...DEFAULTS,
        ...current,
        ...partial,
    };
    next.colorOverrides = { ...(partial.colorOverrides ?? current.colorOverrides ?? {}) };
    next.userPresets = Array.isArray(partial.userPresets)
        ? partial.userPresets
        : [...(current.userPresets || [])];
    return next;
}

function setOrClear(root, prop, value, enabled) {
    if (enabled && value) root.style.setProperty(prop, value);
    else root.style.removeProperty(prop);
}

function applyResolvedVars(root, settings) {
    const colors = resolveColors(settings);
    const overrides = settings.colorOverrides || {};
    const ink = resolveInk(settings);
    const hsl = hexToHsl(colors.accent);
    const width = Math.max(0, Math.min(8, Number(settings.widgetBorderWidth) || 0));
    root.style.setProperty('--accent-h', String(hsl.h));
    root.style.setProperty('--accent-s', `${hsl.s}%`);
    root.style.setProperty('--accent-l', `${hsl.l}%`);
    setOrClear(root, '--bg-canvas', colors.pageBg, !!overrides.pageBg);
    setOrClear(root, '--bg-main', colors.pageBg, !!overrides.pageBg);
    setOrClear(root, '--bg-elevated', colors.widgetBg, !!overrides.widgetBg);
    setOrClear(root, '--bg-panel', colors.widgetBg, !!overrides.widgetBg);
    setOrClear(root, '--bg-sidebar', colors.sidebar, !!overrides.sidebar);
    setOrClear(root, '--text-title', colors.titles, !!overrides.titles);
    setOrClear(root, '--success', colors.done, !!overrides.done);
    setOrClear(root, '--attention', colors.openNext, !!overrides.openNext);
    root.style.setProperty('--home-widget-border-width', `${width}px`);
    setOrClear(root, '--home-widget-border-color', colors.widgetBorder, !!overrides.widgetBorder);
    root.style.setProperty('--on-primary', ink);
    root.style.setProperty('--primary-ink', ink);
}

export function applyAppearanceOverlay(extraOverrides) {
    const merged = {
        ...current,
        colorOverrides: { ...(current.colorOverrides || {}), ...(extraOverrides || {}) },
    };
    applyResolvedVars(document.documentElement, merged);
}

export function applyAppearance(settings) {
    current = mergeSettings(settings);
    const root = document.documentElement;
    root.setAttribute('data-theme', current.theme);
    root.setAttribute('data-density', current.density);
    root.setAttribute('data-radius', current.radius);
    root.setAttribute('data-width', current.width);
    root.setAttribute('data-font', current.font);
    root.setAttribute('data-sidebar', current.sidebar);
    const sidebarToggle = document.getElementById('sidebarToggle');
    if (sidebarToggle) {
        const compact = current.sidebar === 'compact';
        sidebarToggle.setAttribute('aria-expanded', compact ? 'false' : 'true');
        sidebarToggle.setAttribute('aria-label', compact ? 'Expand sidebar' : 'Collapse sidebar');
        sidebarToggle.title = compact ? 'Expand sidebar' : 'Collapse sidebar';
        const collapseLabel = sidebarToggle.querySelector('.nav-label');
        if (collapseLabel) collapseLabel.textContent = compact ? 'Expand' : 'Collapse';
    }
    root.setAttribute('data-today-layout', current.todayLayout || 'split');
    root.setAttribute('data-today-order', current.todayOrder || 'todo,workout,journal');
    root.setAttribute('data-today-todo', current.todayTodo === false ? 'off' : 'on');
    root.setAttribute('data-today-workout', current.todayWorkout === false ? 'off' : 'on');
    root.setAttribute('data-today-journal', current.todayJournal === false ? 'off' : 'on');
    root.setAttribute('data-contrast', current.highContrast ? 'high' : 'normal');
    root.setAttribute('data-motion', current.reducedMotion ? 'reduce' : 'full');
    applyResolvedVars(root, current);
    root.style.setProperty('--journal-font-size', `${current.journalFontSize}px`);
    notifyNativeShell(current);
    listeners.forEach((fn) => {
        try {
            fn(current);
        } catch (e) {
            console.error(e);
        }
    });
}

function notifyNativeShell(settings) {
    const dark = settings.theme !== 'paper';
    try {
        window.webkit?.messageHandlers?.kosistenz?.postMessage({ type: 'theme', dark });
    } catch (_) {
        /* not the native host */
    }
}

export function notifyNativeTab(tab, title) {
    try {
        window.webkit?.messageHandlers?.kosistenz?.postMessage({ type: 'tab', tab, title });
    } catch (_) {
        /* not the native host */
    }
}

export function getAppearance() {
    return {
        ...current,
        colorOverrides: { ...(current.colorOverrides || {}) },
        userPresets: [...(current.userPresets || [])],
    };
}

export function onAppearanceChange(fn) {
    listeners.add(fn);
    return () => listeners.delete(fn);
}

export function snapshotPresetFrom(settings) {
    return {
        baseTheme: settings.theme || 'ocean',
        colors: resolveColors(settings),
        widgetBorderWidth: Math.max(0, Math.min(8, Number(settings.widgetBorderWidth) || 0)),
        inkAuto: settings.inkAuto !== false,
        ink: normalizeHex(settings.ink, ''),
    };
}

export async function persistAppearance(settings) {
    const next = mergeSettings(settings);
    applyAppearance(next);
    writeLocal(next);
    if (typeof eel !== 'undefined' && eel.save_appearance_settings) {
        try {
            const saved = await eel.save_appearance_settings(next)();
            applyAppearance(saved);
            writeLocal(saved);
            return saved;
        } catch (e) {
            console.warn('Could not persist appearance to disk', e);
        }
    }
    return getAppearance();
}

export async function resetAppearance() {
    const kept = [...(current.userPresets || [])];
    const next = { ...DEFAULTS, colorOverrides: {}, userPresets: kept, activePresetId: '' };
    applyAppearance(next);
    writeLocal(next);
    if (typeof eel !== 'undefined' && eel.reset_appearance_settings) {
        try {
            const saved = await eel.reset_appearance_settings()();
            applyAppearance(saved);
            writeLocal(saved);
            return saved;
        } catch (e) {
            console.warn(e);
        }
    }
    return getAppearance();
}

export async function initAppearance() {
    const local = mergeSettings(readLocal());
    applyAppearance(local);
    if (typeof eel !== 'undefined' && eel.get_appearance_settings) {
        try {
            const remote = await eel.get_appearance_settings()();
            const merged = mergeSettings({ ...local, ...remote });
            applyAppearance(merged);
            writeLocal(merged);
            return merged;
        } catch (e) {
            console.warn('Could not load appearance from disk', e);
        }
    }
    return getAppearance();
}

File: appearance.py
"""
Appearance settings — theme, typography, layout, and writing preferences.

Stored as JSON next to other Kosistenz data under Application Support.
"""

from __future__ import annotations

import copy
import json
import os
import re
from pathlib import Path
from typing import Any, Dict, List

import eel

COLOR_SLOTS = (
    "pageBg",
    "widgetBg",
    "widgetBorder",
    "titles",
    "accent",
    "done",
    "openNext",
    "sidebar",
)

INK_LIGHT = "#f7fafc"
INK_DARK = "#1a1814"

# Built-in palettes. CSS theme blocks stay in sync; tests check the hexes.
THEME_PALETTES: Dict[str, Dict[str, str]] = {
    "ocean": {
        "pageBg": "#121c26",
        "widgetBg": "#1d2c3b",
        "widgetBorder": "#2c3d4e",
        "titles": "#eef3f7",
        "accent": "#4f8fcf",
        "done": "#5ebb8e",
        "openNext": "#d4a054",
        "sidebar": "#0e1620",
    },
    "midnight": {
        "pageBg": "#111113",
        "widgetBg": "#1f1f23",
        "widgetBorder": "#2a2a30",
        "titles": "#f4f4f5",
        "accent": "#4f8fcf",
        "done": "#5ebb8e",
        "openNext": "#d4a054",
        "sidebar": "#0c0c0e",
    },
    "slate": {
        "pageBg": "#171e2b",
        "widgetBg": "#243044",
        "widgetBorder": "#2c3848",
        "titles": "#eef2f6",
        "accent": "#4f8fcf",
        "done": "#5ebb8e",
        "openNext": "#d4a054",
        "sidebar": "#121824",
    },
    "paper": {
        "pageBg": "#f7f3eb",
        "widgetBg": "#f1ebe0",
        "widgetBorder": "#d8d0c2",
        "titles": "#1b1814",
        "accent": "#4f8fcf",
        "done": "#2f7d57",
        "openNext": "#b5791f",
        "sidebar": "#f3eee4",
    },
    "forest": {
        "pageBg": "#141e1a",
        "widgetBg": "#20312b",
        "widgetBorder": "#2a3c36",
        "titles": "#eef4f0",
        "accent": "#4f8fcf",
        "done": "#6bc49a",
        "openNext": "#d4a054",
        "sidebar": "#101816",
    },
    "dusk": {
        "pageBg": "#1b1824",
        "widgetBg": "#2b2738",
        "widgetBorder": "#3a3548",
        "titles": "#f2eef6",
        "accent": "#4f8fcf",
        "done": "#5ebb8e",
        "openNext": "#d4a054",
        "sidebar": "#16131e",
    },
}

ACCENT_HEX = {
    "sky": "#4f8fcf",
    "teal": "#2a9a8c",
    "amber": "#c8892c",
    "rose": "#c45c6a",
    "violet": "#7a6cb5",
    "lime": "#6a9a6e",
}

DEFAULTS: Dict[str, Any] = {
    "theme": "ocean",
    "accent": "sky",
    "customAccent": "#4F8FCF",
    "font": "system",
    "density": "comfortable",
    "radius": "soft",
    "width": "standard",
    "sidebar": "expanded",
    "todayLayout": "split",
    "todayOrder": "todo,workout,journal",
    "todayTodo": True,
    "todayWorkout": True,
    "todayJournal": True,
    "journalFontSize": 17,
    "timerMinutes": 10,
    "autoFocus": False,
    "reducedMotion": False,
    "highContrast": False,
    "colorOverrides": {},
    "widgetBorderWidth": 1,
    "inkAuto": True,
    "ink": "",
    "activePresetId": "",
    "userPresets": [],
}

ALLOWED = {
    "theme": {"ocean", "midnight", "slate", "paper", "forest", "dusk"},
    "accent": {"sky", "teal", "amber", "rose", "violet", "lime", "custom"},
    "font": {"sans", "serif", "mono", "system"},
    "density": {"comfortable", "compact"},
    "radius": {"sharp", "soft", "round"},
    "width": {"narrow", "standard", "wide"},
    "sidebar": {"expanded", "compact"},
    "todayLayout": {"split", "stack", "columns"},
}

_TODAY_MODULES = ("todo", "workout", "journal")
_HEX_RE = re.compile(r"^#?[0-9a-fA-F]{3}([0-9a-fA-F]{3})?$")
_PRESET_ID_RE = re.compile(r"^up-[a-zA-Z0-9_-]{2,36}$")
_MAX_PRESETS = 32


def _as_today_order(raw: Any) -> str:
    parts: list[str] = []
    if isinstance(raw, str):
        parts = [p.strip() for p in raw.split(",")]
    elif isinstance(raw, (list, tuple)):
        parts = [str(p).strip() for p in raw]
    seen: list[str] = []
    for part in parts:
        if part in _TODAY_MODULES and part not in seen:
            seen.append(part)
    for part in _TODAY_MODULES:
        if part not in seen:
            seen.append(part)
    return ",".join(seen)


def _app_data_dir() -> Path:
    from paths import data_directory

    return data_directory()


def _settings_path() -> Path:
    return _app_data_dir() / "appearance.json"


def _as_bool(raw: Any, default: bool) -> bool:
    if raw is None:
        return default
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, (int, float)):
        return bool(raw)
    if isinstance(raw, str):
        return raw.strip().lower() in {"1", "true", "yes", "on"}
    return default


def _clamp_int(value: Any, lo: int, hi: int, fallback: int) -> int:
    try:
        n = int(value)
    except (TypeError, ValueError):
        return fallback
    return max(lo, min(hi, n))


def _as_hex(raw: Any, fallback: str = "") -> str:
    if not isinstance(raw, str):
        return fallback
    text = raw.strip()
    if not text:
        return fallback
    if not text.startswith("#"):
        text = "#" + text
    if not _HEX_RE.match(text):
        return fallback
    body = text[1:]
    if len(body) == 3:
        body = "".join(ch * 2 for ch in body)
    return "#" + body.lower()


def _srgb_to_linear(channel: float) -> float:
    return channel / 12.92 if channel <= 0.04045 else ((channel + 0.055) / 1.055) ** 2.4


def relative_luminance(hex_color: str) -> float:
    hx = _as_hex(hex_color, "")
    if not hx:
        return 0.0
    r = int(hx[1:3], 16) / 255.0
    g = int(hx[3:5], 16) / 255.0
    b = int(hx[5:7], 16) / 255.0
    return 0.2126 * _srgb_to_linear(r) + 0.7152 * _srgb_to_linear(g) + 0.0722 * _srgb_to_linear(b)


def ink_for_hex(hex_color: str) -> str:
    """Light ink on dark fills, dark ink on light fills."""
    return INK_DARK if relative_luminance(hex_color) > 0.45 else INK_LIGHT


def palette_for(theme: str) -> Dict[str, str]:
    return dict(THEME_PALETTES.get(theme, THEME_PALETTES["ocean"]))


def _sanitize_overrides(raw: Any) -> Dict[str, str]:
    if not isinstance(raw, dict):
        return {}
    out: Dict[str, str] = {}
    for slot in COLOR_SLOTS:
        hx = _as_hex(raw.get(slot), "")
        if hx:
            out[slot] = hx
    return out


def _sanitize_colors(raw: Any, base_theme: str) -> Dict[str, str]:
    out = palette_for(base_theme)
    if isinstance(raw, dict):
        for slot in COLOR_SLOTS:
            hx = _as_hex(raw.get(slot), "")
            if hx:
                out[slot] = hx
    return out


def _sanitize_preset(raw: Any) -> Dict[str, Any] | None:
    if not isinstance(raw, dict):
        return None
    ident = raw.get("id")
    if not isinstance(ident, str) or not _PRESET_ID_RE.match(ident.strip()):
        return None
    name = raw.get("name")
    if not isinstance(name, str):
        return None
    label = " ".join(name.strip().split())
    if not label or len(label) > 40:
        return None
    base = raw.get("baseTheme")
    if not isinstance(base, str) or base not in ALLOWED["theme"]:
        base = "ocean"
    return {
        "id": ident.strip(),
        "name": label,
        "baseTheme": base,
        "colors": _sanitize_colors(raw.get("colors"), base),
        "widgetBorderWidth": _clamp_int(raw.get("widgetBorderWidth"), 0, 8, 1),
        "inkAuto": _as_bool(raw.get("inkAuto"), True),
        "ink": _as_hex(raw.get("ink"), ""),
    }


def _sanitize_presets(raw: Any) -> List[Dict[str, Any]]:
    if not isinstance(raw, list):
        return []
    out: List[Dict[str, Any]] = []
    seen = set()
    for item in raw:
        preset = _sanitize_preset(item)
        if not preset or preset["id"] in seen:
            continue
        seen.add(preset["id"])
        out.append(preset)
        if len(out) >= _MAX_PRESETS:
            break
    return out


def resolve_accent_hex(settings: Dict[str, Any]) -> str:
    overrides = settings.get("colorOverrides") or {}
    if isinstance(overrides, dict):
        hx = _as_hex(overrides.get("accent"), "")
        if hx:
            return hx
    if settings.get("accent") == "custom":
        hx = _as_hex(settings.get("customAccent"), "")
        if hx:
            return hx
    preset = ACCENT_HEX.get(str(settings.get("accent") or "sky"))
    if preset:
        return preset
    return palette_for(str(settings.get("theme") or "ocean"))["accent"]


def resolve_colors(settings: Dict[str, Any]) -> Dict[str, str]:
    theme = str(settings.get("theme") or "ocean")
    out = palette_for(theme)
    overrides = settings.get("colorOverrides") if isinstance(settings.get("colorOverrides"), dict) else {}
    for slot in COLOR_SLOTS:
        if slot == "accent":
            continue
        hx = _as_hex(overrides.get(slot), "")
        if hx:
            out[slot] = hx
    out["accent"] = resolve_accent_hex(settings)
    return out


def resolve_ink(settings: Dict[str, Any]) -> str:
    if _as_bool(settings.get("inkAuto"), True):
        return ink_for_hex(resolve_accent_hex(settings))
    hx = _as_hex(settings.get("ink"), "")
    return hx or ink_for_hex(resolve_accent_hex(settings))


def _sanitize(raw: Any) -> Dict[str, Any]:
    if not isinstance(raw, dict):
        return copy.deepcopy(DEFAULTS)
    out = copy.deepcopy(DEFAULTS)
    for key, allowed in ALLOWED.items():
        val = raw.get(key)
        if isinstance(val, str) and val in allowed:
            out[key] = val
    custom = raw.get("customAccent")
    if isinstance(custom, str) and len(custom) <= 16:
        hx = _as_hex(custom.strip(), "")
        out["customAccent"] = hx or DEFAULTS["customAccent"]
    out["journalFontSize"] = _clamp_int(raw.get("journalFontSize"), 14, 22, DEFAULTS["journalFontSize"])
    out["timerMinutes"] = _clamp_int(raw.get("timerMinutes"), 5, 30, DEFAULTS["timerMinutes"])
    out["autoFocus"] = _as_bool(raw.get("autoFocus"), False)
    out["reducedMotion"] = _as_bool(raw.get("reducedMotion"), False)
    out["highContrast"] = _as_bool(raw.get("highContrast"), False)
    out["todayTodo"] = _as_bool(raw.get("todayTodo"), True)
    out["todayWorkout"] = _as_bool(raw.get("todayWorkout"), True)
    out["todayJournal"] = _as_bool(raw.get("todayJournal"), True)
    out["todayOrder"] = _as_today_order(raw.get("todayOrder"))
    out["colorOverrides"] = _sanitize_overrides(raw.get("colorOverrides"))
    out["widgetBorderWidth"] = _clamp_int(raw.get("widgetBorderWidth"), 0, 8, DEFAULTS["widgetBorderWidth"])
    out["inkAuto"] = _as_bool(raw.get("inkAuto"), True)
    out["ink"] = _as_hex(raw.get("ink"), "")
    out["userPresets"] = _sanitize_presets(raw.get("userPresets"))
    preset_id = raw.get("activePresetId")
    if isinstance(preset_id, str) and any(p["id"] == preset_id.strip() for p in out["userPresets"]):
        out["activePresetId"] = preset_id.strip()
    else:
        out["activePresetId"] = ""
    return out


@eel.expose
def get_appearance_settings() -> Dict[str, Any]:
    path = _settings_path()
    if not path.exists():
        return copy.deepcopy(DEFAULTS)
    try:
        with open(path, "r", encoding="utf-8") as f:
            return _sanitize(json.load(f))
    except (OSError, json.JSONDecodeError):
        return copy.deepcopy(DEFAULTS)


@eel.expose
def save_appearance_settings(settings: Dict[str, Any]) -> Dict[str, Any]:
    cleaned = _sanitize(settings)
    path = _settings_path()
    tmp = str(path) + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(cleaned, f, indent=2)
    os.replace(tmp, path)
    return cleaned


@eel.expose
def reset_appearance_settings() -> Dict[str, Any]:
    current = get_appearance_settings()
    fresh = copy.deepcopy(DEFAULTS)
    fresh["userPresets"] = current.get("userPresets") or []
    return save_appearance_settings(fresh)
